### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by Speiger** [BuiltBrokenModding/AI-Improvements#17]
o/ 
Used your mod in 1.16.5.
AI improvements sadly does not do what it offers.
The mod causes mobs to react only if you are like 1 block away from mobs, otherwise they are frozen.
And if they react they act terribly weird.
On top of that on our weak server we had usually 30ms lag, the moment we fought against mobs the server would spike to 150ms of lag fighting a single zombie it would go up to 100ms.
Letting a player join would also drive up the lag for multiple minutes to +300ms of lag.

After uninstalling the mod, all lag problems we had regarding mobs were basically gone.
Players can join like normal without killing the server on accident.

So TL:DR: This mod, breaks normal AI behavior and causes massive lag.
Sadly due to mixins its not really possible with Spark to detect whats causing it.

I can not recommend this mod unless you want to lag the server really badly.

**Comment by DarkGuardsman** [comment]
If you could provide performance logs that would be useful. That said AI improvements only changes the look calculation math on the AIs. Which is often only used for idle animation and looking at targets. It does nothing to change their AI logic itself. So if they are frozen this may not be a result of the mod but TPS dipping too low. 

That said since you said you didn't want support on our curse page I'm closing this ticket. You can reopen if you provide logs and want support.

> AI improvements sadly does not do what it offers.

> I can not recommend this mod unless you want to lag the server really badly.

Issue tickets are not review sections XD

**Comment by Speiger** [comment]
Well it offers a "Improved AI" not a AI where you can walk by and mobs are not reacting to you.
Some players on my servers walked by creepers (not running, walking) and they didn't even notice them until they were already so far away that the AI stopped again because to far away.

I call this not a improvement, the lag fest is different thing with just stating a fact.
Not a review, just stating facts.

**Comment by Speiger** [comment]
Anyways yeah i have no issues with the issue being closed, because this was more to signal boost other issue reporters, which worked, even if i had to get a bit more drastic.

**Comment by DarkGuardsman** [comment]
> Not a review, just stating facts.

I have yet to see you backup anything to make it a fact. I'm willing to look into issues if I have data but so far this is reading more & more like a rant based on speculation. Could this mod be causing lag? Sure, but I'm having trouble seeing how it is breaking mobs. This is why it is really important to provide things like logs, performance data, mod lists, and video examples. As it is possible another mod is having a bad reaction with our mod. Especially given how easy it is to mixen changes to AIs now. We personally don't use mixens but AI task replacement. Not all mods react well to this if they expect a task to be run or exist for their own edits. 

> Some players on my servers walked by creepers (not running, walking) and they didn't even notice

This sounds like the targeting AI failing. We do not edit the targeting or attack AI. Only thing I can think of is your server is lagging so bad the player position is updating quicker then the server AI tick. This is possible as client controls position priority over the server. 

You can look through our list of changes... https://github.com/BuiltBrokenModding/AI-Improvements/blob/1.16/src/main/java/com/builtbroken/ai/improvements/modifier/ModifierSystem.java#L42

For combat mobs the only two AI tasks that are overriden are LookAtGoal and LookRandomlyGoal. Both of which are server side handling of head rotation only. If removed they don't even break the AI logic at all. Since none of the mobs actually check if they are aiming before attacking.

> this was more to signal boost other issue reporters, which worked, even if i had to get a bit more drastic.

No this was a rant, you didn't need to signal boost. Sure it pissed me off enough I actually bothered to respond... but it still doesn't change my lack of time to do much with the mod. We currently only target game crashing bug fixes. Performance hits we just recommend people change the configs or remove the mod. It sucks but other things take priority such as life, job, and keeping my depression in check. 

Also reacting this way to mod developers just annoys us heavily and demotivates us from wanting to work on the projects. Hell it is the reason I rarely interact with the community these days, have every last github repo muted, emails disable, etc. Only checking things manually once a week or on the rare moments I need to fix github actions on other projects.

**Comment by Speiger** [comment]
> I have yet to see you backup anything to make it a fact. I'm willing to look into issues if I have data but so far this is reading more & more like a rant based on speculation. Could this mod be causing lag? Sure, but I'm having trouble seeing how it is breaking mobs. This is why it is really important to provide things like logs, performance data, mod lists, and video examples. As it is possible another mod is having a bad reaction with our mod. Especially given how easy it is to mixen changes to AIs now. We personally don't use mixens but AI task replacement. Not all mods react well to this if they expect a task to be run or exist for their own edits.

Well very simple: Yeah removing the Look around actually reduces the effectiveness of the AI, because line of sight is a thing so if you remove the "Look around" task the mob can actually not see you. (EntitySenses.canSee()) function should help you with that.
The less reaction could be also that the AI was on cooldown and the extra lag made it so much slower that yeah...

So the AI breaks because it no longer looks around and the max sense distance (without view) is 0 blocks for Mobs until it gets changed (FollowRangeAttribute)

My example that we had may be a extreme case, but AI has to look at you to actually attack you otherwise it would be really bad.

> No this was a rant, you didn't need to signal boost. Sure it pissed me off enough I actually bothered to respond... but it still doesn't change my lack of time to do much with the mod. We currently only target game crashing bug fixes. Performance hits we just recommend people change the configs or remove the mod. It sucks but other things take priority such as life, job, and keeping my depression in check.

Honestly at this point we are talking opinions, simple solution: One side has to stop answering. I think i am going to start after this comment on that suggestions, and if you react to a frustration comment (on curseforge) this badly, then i can understand that, but i would suggest to look why that is triggered in the first place. I went through that myself.

And without this frustration comment you would maybe have not seen this for a couple other months letting get the reputation even worse for the reporters. (It's not like people who report bugs like do not talk to other people about mods, in fact in the 4 days i waited i already warned a couple people about this mod, starting the 3rd-4th day of no respond time)

Small edit: Discord Bots for issue reporting exist, dunno how to setup one though

**Comment by DarkGuardsman** [comment]
> Yeah removing the Look around actually reduces the effectiveness of the AI

We don't remove it but replace it with different math. You have to enable the remove config manually. Which only breaks animation and not AI functionality. 

Also we have configs to disable replacing the math.... for the exact reasons of mod compatibility. Which is what I think the problem is here rather than the mod itself is tanking your server. We don't test against other AI adjustment, improvement, or tweaking mods. So it would not surprise me if one eventually conflicted with our changes. 

> line of sight is a thing so if you remove the "Look around" task the mob can actually not see you 

LookHelper and line of sight checks are two different segments of code. LookHelper handles head rotation when aiming at a block position or entity vector. Line of sight is a raytrace fired from one position to the next ignoring head rotation. I will double check the code to confirm this has not changed. However, I'm rather sure mojang still doesn't check for head rotation when doing line of sight as the skeletons still shoot backwards in 1.16 vanilla. 

> Honestly at this point we are talking opinions, simple solution: One side has to stop answering

Cool locking this thread. Only reason I've reacted this negative is due to how badly you are making guesses on how things function. As the more misinformation that exists the worse people take guesses and blame the mod for random things. Its already bad enough I get direct messages for every last AI problem people spot. I don't need people spreading rumors about bugs that exist that may have nothing to do with the mod. Want this opinion to change... then open a proper issue ticket with data so me or one of the contributors can look into it.

> And without this frustration comment you would maybe have not seen this for a couple other months letting get the reputation even worse for the reporters

As I told you on curse, I did see these issues but decided to not respond. I do this often as it can take hours to read and respond to everything each day. Instead I pick out a few that require responses and leave the rest alone. It sucks for people honestly trying to help but has to be done to keep from getting overwhelmed. 

> already warned a couple people about this mod

Good, maybe people will think twice about installing a mod just because it says 'improvement' in the title. Rather than looking into the features offered, warnings that it is not a magic cure, and that I'm not a god that has fixed every last problem with AIs. I honestly don't need the downloads and get depressed each time I look at them. As a lot of people install this mod without ever checking if it works. After 1.12 the performance improvements of the math changes are less than 1%. Only reason it even gets maintained is for those rare edge cases where it fixes bad servers. Such as one my friend runs that has massive sheep farms which tank TPS due to idle look animation. 

> Discord Bots for issue reporting exist

We had one at one point and disabled it... they are not as nice as they sound and often just frustrate people with random pings at 3am. Honestly if our team becomes active again we might consider a full bot for it.

**Comment by DarkGuardsman** [comment]
Checked the code... raytrace for line of sight still doesn't consider head rotation or aim. I also checked zombie, creeper, and skeleton AI tasks. All though the skeleton has an improved aiming animation it still doesn't check for facing. Zombie also doesn't even check for line of sight... only distance and pathfinding before attacking. Creeper does check for line of sight but doesn't check for aiming.
